docs/architecture/adr-003-stateful-precompiles.md

```diff
@@ -0,0 +1,192 @@
+	stateDB, ok := evm.StateDB.(ExtStateDB)
+	if !ok {
+		return nil, errors.New("not run in ethermint")
+	}
+
+	// parse input
+	methodID := input[:4]
+	if bytes.Equal(methodID, MintMethod.ID) {
+		if readonly {
+			return nil, errors.New("the method is not readonly")
+		}
+		args, err := MintMethod.Inputs.Unpack(input[4:])
+		if err != nil {
+			return nil, errors.New("fail to unpack input arguments")
+		}
+		recipient := args[0].(common.Address)
+		amount := args[1].(*big.Int)
+		if amount.Sign() <= 0 {
+			return nil, errors.New("invalid amount")
+		}
+
+		if _, ok := bc.balances[caller]; !ok {
+			bc.balances[caller] = make(map[common.Address]*Balance)
+		}
+		balances := bc.balances[caller]
+		if balance, ok := balances[recipient]; ok {
+			balance.DirtyAmount = new(big.Int).Add(balance.DirtyAmount, amount)
+		} else {
+			// query original amount
+			addr := sdk.AccAddress(recipient.Bytes())
+			originAmount := bc.bankKeeper.GetBalance(bc.ctx, addr, EVMDenom(caller)).Amount.BigInt()
+			dirtyAmount := new(big.Int).Add(originAmount, amount)
+			balances[recipient] = &Balance{
+				OriginAmount: originAmount,
+				DirtyAmount:  dirtyAmount,
+			}
+		}
+		stateDB.AppendJournalEntry(bankMintChange{bc: bc, caller: caller, recipient: recipient, amount: amount})
+  } else if bytes.Equal(methodID, BalanceOfMethod.ID) {
+		args, err := BalanceOfMethod.Inputs.Unpack(input[4:])
+		if err != nil {
+			return nil, errors.New("fail to unpack input arguments")
+		}
+		token := args[0].(common.Address)
+		addr := args[1].(common.Address)
+		if balances, ok := bc.balances[token]; ok {
+			if balance, ok := balances[addr]; ok {
+				return BalanceOfMethod.Outputs.Pack(balance.DirtyAmount)
+			}
+		}
+		// query from storage
+		amount := bc.bankKeeper.GetBalance(bc.ctx, sdk.AccAddress(addr.Bytes()), EVMDenom(token)).Amount.BigInt()
+		return BalanceOfMethod.Outputs.Pack(amount)
+	} else {
+		return nil, errors.New("unknown method")
+	}
+  return nil, nil
+}
+
+func (bc *BankContract) Commit(ctx sdk.Context) error {
+  // Write the dirty balances through bc.bankKeeper
+}
+
+type bankMintChange struct {
+	bc        *BankContract
+	caller    common.Address
+	recipient common.Address
+	amount    *big.Int
+}
+
+func (ch bankMintChange) Revert(*statedb.StateDB) {
+	balance := ch.bc.balances[ch.caller][ch.recipient]
+	balance.DirtyAmount = new(big.Int).Sub(balance.DirtyAmount, ch.amount)
+}
+
+func (ch bankMintChange) Dirtied() *common.Address {
+	return nil
+}
+```
+
+## Consequences
+
+> This section describes the resulting context, after applying the decision. All consequences should be listed here, not just the "positive" ones. A particular decision may have positive, negative, and neutral consequences, but all of them affect the team and project in the future.
+
+### Backwards Compatibility
+
+- State machine breaking
+- Need to patch `go-ethereum`
```

Comment:
this is a no-go, we don't want to maintain a geth fork

Reply:
This is very solvable via shadowing Call, CallCode, CallDelegate etc.

Reply:
> This is very solvable via shadowing Call, CallCode, CallDelegate etc.

Yeah, this is what I had in mind. Ideally there are multiple EVM implementations that support the precompiles